### PR TITLE
feat: use middleware for .io rewrites

### PR DESCRIPTION
**Pull request by dstaley**
## "Ready for Review" checklist

- [ ] The Vercel preview link has been added to this PR's description
- [ ] The Asana task has been added to this PR's description
- [ ] This PR's link has been added to the "📐 Pull Request" field on the Asana task
- [ ] This Vercel preview link has been added to the "🔍 Preview" field on the Asana task
- [ ] The description template has been filled in below

---

## Relevant links

- [Preview link](https://dev-portal-git-BRANCH_NAME-hashicorp.vercel.app/PATH_TO_VIEW) 🔎
- [Asana task](url) 🎟️

## What

Simplifies the rewrite logic by shifting it to `pages/_middleware.ts`

## Why

To support more robust e2e testing, and to eventually support the easier removal of .io rewrites.

## How

In `pages/_middleware.ts`, we examine every incoming request to determine if the request is product-specific. If so, we rewrite all URLs (minus static assets) to the respective `/_proxied-dot-io` version.

## Testing

- Go to preview URL: https://dev-portal-5xxqaf5l9-hashicorp.vercel.app/
- Change to a product
- Ensure homepage renders without issue
- Repeat for remaining products
- Switch to Nomad
- Ensure this URL returns a JSON file: https://dev-portal-5xxqaf5l9-hashicorp.vercel.app/data/vault/nomad-cluster-role.json
- 

## Anything else?

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201980045498143

### CODE REVIEW

File: src/pages/_middleware.ts [hunk 1]
```diff
@@ -77,6 +91,19 @@ export function middleware(req: NextRequest, ev: NextFetchEvent) {
     return NextResponse.redirect(url, permanent ? 308 : 307)
   }
 
+  // Rewrite .io pages
```
Comment (BRKalow) [inline src/pages/_middleware.ts:94]: We might want to change the precedence in `determineProductSlug` with this change. I think the hostname check should take priority, before the cookie.

It's likely not a huge deal, as I would need to have an `io_preview` cookie set on packer.io that is NOT packer, for example, but seems like the correct ordering.

File: build-libs/rewrites.js [hunk 2]
```diff
@@ -1,85 +0,0 @@
-//@ts-check
```
Comment (BRKalow) [inline build-libs/rewrites.js:1]: So excited to see this simplified!

Comment (BRKalow) [inline build-libs/rewrites.js:1]: So excited to see this simplified!

**Comment by dstaley** [comment]
This has probably diverged enough from the base branch that it should probably be re-created from scratch if this is functionality that's still desirable. I'll defer to you and @BRKalow on whether it makes sense to implement this approach given how the middleware currently works in dev-portal.

**Comment by BRKalow** [comment]
Let's close this out for now 👍

**Pull request closed without merging**